src/Accounts/Authentication/Utilities/FrequencyService.cs

```diff
@@ -0,0 +1,205 @@
+                    business();
+                }
+            }
+            else if (_frequencies.ContainsKey(featureName))
+            {
+                if (_clock.IsDue(_frequencies[featureName].LastCheckTime, _frequencies[featureName].Frequency) && businessCheck())
+                {
+                    _frequencies[featureName].LastCheckTime = DateTime.Now;
+                    business();
+                    Save();
+                }
+            }
+            else
+            {
+                throw new ArgumentException($"Feature name '{featureName}' not found in FrequencyService");
+            }
+
+        }
+        /// <summary>
+        /// This method is used for testing purposes only. It allows the caller to specify a fakeNow time to use for the check.
+        /// </summary>
+        internal void Check(string featureName, Func<bool> businessCheck, Action business, DateTime fakeNow)
+        {
+            if (_perPSSessionRegistry != null && _perPSSessionRegistry.ContainsKey(featureName))
+            {
+                if (_perPSSessionRegistry[featureName] == false && businessCheck())
+                {
+                    _perPSSessionRegistry[featureName] = true;
+                    business();
+                }
+            }
+            else if (_frequencies.ContainsKey(featureName))
+            {
+                if (_clock.IsDue(_frequencies[featureName].LastCheckTime, _frequencies[featureName].Frequency) && businessCheck())
+                {
+                    _frequencies[featureName].LastCheckTime = fakeNow;
+                    business();
+                    Save();
+                }
+            }
+            else
+            {
+                throw new ArgumentException($"Feature name '{featureName}' not found in FrequencyService");
+            }
+
+        }
+
+        public void Register(string featureName, TimeSpan frequency)
+        {
+            if (!_frequencies.ContainsKey(featureName))
+            {
+                _frequencies.Add(featureName, new FrequencyInfo(frequency, DateTime.MinValue));
+                Save();
+            }
+            if (_frequencies.ContainsKey(featureName) && _frequencies[featureName].Frequency != frequency)
+            {
+                throw new ArgumentException($"Feature name '{featureName}' already exists in FrequencyService with a different frequency!");
+            }
+        }
+
+        public void RegisterInSession(string featureName)
+        {
+            if (_perPSSessionRegistry == null)
+            {
+                _perPSSessionRegistry = new Dictionary<string, bool>();
+            }
+            if (!_perPSSessionRegistry.ContainsKey(featureName))
+            {
+                _perPSSessionRegistry.Add(featureName, false);
+            }
+        }
+
+        public void Save()
+        {
+            string json = JsonConvert.SerializeObject(_frequencies);
+            _dataStore.WriteFile(_filePath, json);
```

Comment:
What happens if two frequency services write file at the same time?

Reply:
Some exceptions could happen:

- More than 1 Azure PowerShell Session update the frequency file => the last one will overwrite the former ones. => the frequency task in former ones may have wrong "due date" next time => expected.
- Failed to read the frequency file => treat all frequency tasks the first time => expected.
- Failed to parse the frequency file => treat all frequency tasks the first time => expected.